### PR TITLE
feat: add wrapper-based API testing approach

### LINKED ISSUE #241: [Feature]: Add Decorator-Based API Testing Support for Enhanced
### Current Situation/Problem

Currently, itdoc provides excellent test-driven API documentation generation through its function-based approach using `describeAPI()` and `itDoc()`. However, this approach requires developers to write verbose boilerplate code and manually integrate with test frameworks

### Proposed Feature

Introduce **decorator-based API documentation support** that allows developers to add documentation decorators to existing E2E tests in their preferred test frameworks, providing a seamless way to generate API documentation from existing test suites:

```typescript
// Proposed decorator-based approach - enhancing existing E2E tests
import { ApiDoc, DocumentEndpoint } from 'itdoc/decorators'
import request from 'supertest'
import { app } from '../src/app'

describe('User Authentication API', () => {
    
    @ApiDoc({
        method: "POST",
        path: "/signup",
        summary: "User Registration API",
        description: "Register a new user with username and password",
        tag: "Auth"
    })
    it('should register user successfully', async () => {
        const response = await request(app)
            .post('/signup')
            .send({
                username: 'testuser',
                password: 'password123'
            })
            .expect(201)
        
        expect(response.body).toEqual({
            message: 'User registered successfully',
            userId: expect.any(String)
        })
    })

    @ApiDoc({
        method: "POST",
        path: "/signup",
        summary: "User Registration API - Validation Error",
        description: "Returns error when username is missing",
        tag: "Auth"
    })
    it('should fail when username is missing', async () => {
        const response = await request(app)
            .post('/signup')
            .send({
                password: 'password123'
            })
            .expect(400)
        
        expect(response.body).toEqual({
            error: 'Username is required'
        })
    })
})

// Works with any test framework
describe('Product API', () => {
    
    @DocumentEndpoint({
        method: "GET",
        path: "/products/:id",
        summary: "Get Product by ID",
        tag: "Products"
    })
    test('should return product details', async () => {
        // Existing Jest/Mocha/Vitest test code
        const response = await request(app)
            .get('/products/123')
            .expect(200)
        
        expect(response.body.id).toBe(123)
        expect(response.body.name).toBeDefined()
    })
})
```

**Benefits of decorator-based approach:**
1. **Non-intrusive integration**: Add documentation to existing E2E tests without changing test logic
2. **Framework agnostic**: Works seamlessly with any test framework (Jest, Mocha, Vitest, etc.)
3. **Minimal learning curve**: Developers can keep using their familiar testing patterns
4. **Better IDE support**: Enhanced autocomplete, IntelliSense, and refactoring capabilities
5. **Gradual adoption**: Teams can incrementally add documentation decorators to existing tests
6. **Automatic documentation generation**: Extract API specs directly from real, working tests
7. **Test-driven documentation**: Documentation is always in sync with actual test behavior

### Alternatives (Optional)

_No response_

### Additional Information

_No response_

### LINKED ISSUE #247: [Feature]: Please Support Multipart
Currently, itdoc does not seem to have native support for documenting APIs that use multipart/form-data (e.g., file uploads combined with JSON fields).

It would be useful if itdoc could:

Allow defining multipart request bodies with mixed types (files + text fields).

Generate examples in the final documentation that clearly show how clients should construct multipart requests.

Optionally, support multipart responses if applicable.

This is a fairly common use case (especially for file uploads), and having first-class support in itdoc would make the generated API docs much clearer and more complete.

### PR DESCRIPTION
**Pull request by json-choi**
# High-Order Function Wrapper for Automatic API Documentation Generation

## Overview
Implement high-order function wrapper for Jest/Mocha to capture HTTP requests/responses automatically for API documentation generation.

### Minimal Code Changes
```typescript
// Before
import request from "supertest"
it("should create user", async () => {
  const response = await request(app).post("/users").send({ name: "John" })
  expect(response.status).toBe(201)
})

// After  
import { wrapTest, createClient } from "itdoc/wrappers"
const apiTest = wrapTest(it)
apiTest("should create user", async () => {
  const response = await createClient.supertest(app).post("/users").send({ name: "John" })
  expect(response.status).toBe(201)
  // Automatic request/response capture!
})
```

### Key Components
- **wrapTest()**: Wraps Jest/Mocha 'it' function with capture context
- **createClient**: Multi-adapter HTTP client (Supertest, Axios, Fetch)
- **CaptureContext**: AsyncLocalStorage for thread-safe context management
- **withMeta()**: API documentation metadata support

### Advanced Features
- **File Upload Support**: Multipart/form-data with proper form field capture (fixes #247)
- **Multiple HTTP Clients**: Supertest, Axios, Fetch adapters
- **Framework Agnostic**: Full Jest/Mocha support
- **Thread-Safe**: AsyncLocalStorage for concurrent test execution
- **Backward Compatible**: Coexists with existing itDoc DSL

### Core Files Added
- `lib/wrappers/wrapTest.ts` - Main wrapper function
- `lib/wrappers/core/CaptureContext.ts` - Context management
- `lib/wrappers/adapters/` - HTTP client adapters
- `lib/wrappers/types.ts` - TypeScript definitions
- `examples/express-ts/src/__tests__/*.wrapper.test.ts` - Test examples

### Test Coverage
- Unit tests for CaptureContext, interceptedRequest, wrapTest
- Integration tests for complete workflow
- Real-world examples: User, Product, Upload APIs
- Jest and Mocha compatibility tests

#### Basic Usage
```typescript
const apiTest = wrapTest(it)
const request = createClient.supertest(app)

apiTest("should get users", async () => {
  const response = await request.get("/users")
  expect(response.status).toBe(200)
})
```

#### With Metadata
```typescript
apiTest.withMeta({
  summary: "Create User",
  tags: ["Users", "Authentication"],
  description: "Register a new user account"
})("should create user", async () => {
  const response = await request.post("/users").send({ name: "John" })
  expect(response.status).toBe(201)
})
```

#### File Upload
```typescript
apiTest("should upload file", async () => {
  const response = await request
    .post("/api/upload")
    .field("title", "Document")
    .attach("file", "path/to/file.pdf")
  
  expect(response.status).toBe(201)
})
```

#### Multiple HTTP Clients
```typescript
// Supertest (Express apps)
const response = await createClient.supertest(app).get("/users")

// Axios (External APIs)  
const response = await createClient.axios({ baseURL: "https://api.example.com" }).get("/users")

// Fetch (Modern approach)
const response = await createClient.fetch({ baseURL: "https://api.example.com" }).get("/users")
```

### From Existing Supertest
```typescript
// 1. Change import
- import request from "supertest"
+ import { wrapTest, createClient } from "itdoc/wrappers"

// 2. Wrap it function
+ const apiTest = wrapTest(it)

// 3. Use createClient
- const response = await request(app).post("/users")
+ const response = await createClient.supertest(app).post("/users")

// 4. Change it to apiTest
- it("should create user", async () => {
+ apiTest("should create user", async () => {
```

## Benefits
- **Zero Breaking Changes**: Additive feature only
- **Minimal Migration**: 1-2 line changes in most cases
- **Automatic Capture**: No manual request/response setup
- **Type Safe**: Full TypeScript support
- **Framework Agnostic**: Works with Jest and Mocha
- **Thread Safe**: Concurrent test execution support

## Known Issues & TODOs
- Needs comprehensive code review
- Requires additional edge case testing  
- Documentation needs improvement
- Type definitions may need refinement
- ESLint warnings to be addressed
- AsyncLocalStorage context timing requires validation

This is a draft implementation for issue #241. Further review and testing required before production use.

Related: #241, #247

## Summary by CodeRabbit

- New Features
  - Introduced wrapper-based API testing with wrapTest and createClient.
  - Added HTTP client adapters: Supertest, Axios, and Fetch.
  - Expanded Express example with file upload endpoints and improved product service reset.
- Documentation
  - Added comprehensive guides and examples for wrappers and API testing.
- Tests
  - Added extensive unit and integration tests for wrappers, adapters, and example APIs (Jest/Mocha).
- Chores
  - Updated test configs/paths and dependencies.
  - Added project configuration and gitignore updates.

### CODE REVIEW

File: lib/wrappers/adapters/axios/AxiosAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+export interface AxiosAdapterConfig {
+    baseURL?: string
+    timeout?: number
+    headers?: Record<string, string>
+}
+
+export class AxiosAdapter implements HttpAdapter<AxiosAdapterConfig> {
+    public create(config: AxiosAdapterConfig): HttpClient {
+        const axios = require("axios")
```
Comment (PENEKhun) [inline lib/wrappers/adapters/axios/AxiosAdapter.ts:29]: [P1] ESM consumers cannot create the Axios adapter. The package exposes `build/index.js` through `exports.import`, but tsup leaves this call as a dynamic `require("axios")`. I reproduced `Dynamic require of "axios" is not supported` with `import { createClient } from "itdoc"; createClient.axios(...)`. Use an ESM-safe loading strategy and test the actual built import entry point.

File: lib/wrappers/adapters/fetch/FetchAdapter.ts [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        }
+
+        if (typeof file === "string") {
+            const fs = require("fs")
+            this.formDataObj.append(field, fs.createReadStream(file), filename || file)
+        } else {
+            this.formDataObj.append(field, file, filename || "file")
+        }
+
+        this.init.body = this.formDataObj as any
```
Comment (PENEKhun) [inline lib/wrappers/adapters/fetch/FetchAdapter.ts:123]: [P1] An npm `form-data` instance is not a WHATWG `BodyInit` for Node's native `fetch`. Even after replacing the preceding dynamic `require` with an ESM import, this value is sent as the literal string `[object FormData]`, not multipart bytes; I confirmed this with a local HTTP server. Use Node 20's native `FormData` with `Blob`/`File`, or another fetch-compatible multipart encoder, so `createClient.fetch(...).attach()` actually uploads the file.

File: lib/wrappers/index.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+/**
+ * Wrapper-based API testing module
+ *
+ * This module provides a high-order function approach to automatically capture
+ * HTTP requests and responses from your tests and generate API documentation.
+ * @example Basic usage
+ * ```typescript
+ * import { wrapTest, createClient } from 'itdoc/wrappers'
```
Comment (PENEKhun) [inline lib/wrappers/index.ts:24]: [P1] This documented public import path does not exist in the package. `package.json` exports only `"."`, and tsup builds only the root entry, so `import("itdoc/wrappers")` fails with `ERR_PACKAGE_PATH_NOT_EXPORTED`. Since the root already exports `wrapTest` and `createClient`, either change the example to `from "itdoc"` or add a real `./wrappers` build/export entry.

File: lib/wrappers/__tests__/CaptureContext.test.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { describe, it } from "mocha"
+import { expect } from "chai"
+import { CaptureContext } from "../core/CaptureContext"
+
+describe("CaptureContext", () => {
```
Comment (PENEKhun) [inline lib/wrappers/__tests__/CaptureContext.test.ts:21]: [P2] These new tests are not part of the default CI test run. The root Mocha config matches only `lib/__tests__/**/*.test.ts`, so `pnpm test:unit` ran only the existing 37 tests; `lib/wrappers/__tests__` ran only when I supplied a manual glob. Include this directory in the default spec and add tests against the built ESM Axios/Fetch entry points so the blockers above cannot be missed again.

**Review by PENEKhun — CHANGES_REQUESTED** [review]
I verified the PR head with `pnpm build`, the default unit/e2e suites, and a manual run of the wrapper tests. Existing tests pass, but the core API fails when Axios and Fetch multipart are invoked through the published ESM entry point, and the documented public import path does not resolve. The new wrapper test directory is also omitted from the default test command, and the PR currently conflicts with `develop`. Please address the blockers below and rebase/clean up against the latest base before re-review.
